### ISSUE TITLE
Tool support for browser backend

### ISSUE THREAD

**Issue by thehunmonkgroup** [mmabrouk/chatgpt-wrapper#300]
Tools are now rolling out to Plus users, so I took a quick look at how hard it would be to integrate them into the browser backend.

This is the basics:

```diff
--- a/chatgpt_wrapper/backends/browser/backend.py
+++ b/chatgpt_wrapper/backends/browser/backend.py
@@ -494,6 +494,9 @@ class BrowserBackend(Backend):
             "conversation_id": self.conversation_id,
             "parent_message_id": self.parent_message_id,
             "action": "next",
+            "plugin_ids": [
+                "plugin-d1d6eb04-3375-40aa-940a-c2fc57ce0f51"
+            ],
         }

         code = (
diff --git a/chatgpt_wrapper/plugins/provider_chatgpt_browser.py b/chatgpt_wrapper/plugins/provider_chatgpt_browser.py
index 4002952..2ac32d5 100644
--- a/chatgpt_wrapper/plugins/provider_chatgpt_browser.py
+++ b/chatgpt_wrapper/plugins/provider_chatgpt_browser.py
@@ -94,6 +94,8 @@ class ProviderChatgptBrowser(Provider):
                 },
                 'gpt-4': {
                 },
+                'gpt-4-plugins': {
+                },
             }
         }
```

So there's a new model type `gpt-4-plugins`, and a new parameter `plugin_ids` that gets passed a list of plugin IDs, and that will activate the tool for the request.

The streaming output looked a tad funky, but the final output is actually quite nice:

```
gpt-4-plugins 2> /chat
User:

how far is it from the sun to the earth?

Assistant:

{ "input": "distance from sun to earth" }

Tool:

Query: "distance from sun to earth"

Assumption: Assuming "sun" is referring to stars To use as a word set assumption=*C.sun-_*EnglishWord-

Input interpretation: Sun | distance from Earth

Current result: 1.01 au (astronomical units)

Unit conversions: 1.511×10^8 km (kilometers)

1.511×10^11 meters

93.92 million miles

Comparisons as distance: ≈ ( 0.011 ≈ 1/87 ) × smallest distance from the Sun to the heliosheath ( 79 to 100 au )

≈ 1.01 × mean Earth-Sun distance ( 1.0000010178 au )

Corresponding quantities: Light travel time t in vacuum from t = x/c: | 8.4 minutes

Light travel time t in an optical fiber t = 1.48x/c: | 12 minutes

Solar radiation pressure from P = L_☉/(c4πr^2): | 4.4 μPa (micropascals) | (assuming solar constant ≈ 1.36 kW/m^2)

Wolfram|Alpha website result for "distance from sun to earth": https://www.wolframalpha.com/input?i=distance+from+sun+to+earth

Assistant:

The average distance from the Sun to the Earth is approximately:

 • 1.01 astronomical units (AU)
 • 1.511×10^8 kilometers
 • 93.92 million miles

This distance is also equivalent to the light travel time of about 8.4 minutes in a vacuum.
```

It would be trivial to add a new `browser.plugins` config option, which could take a list of plugin IDs. Kind of a pain to have to look up the plugin IDs manually, though. Plus there would be no mechanism to enable/disable plugins on the fly. It would take some investigation to make the solution more elegant.

**Comment by thehunmonkgroup** [comment]
Can anyone else confirm their Wolfram Alpha plugin has the following plugin ID? `plugin-d1d6eb04-3375-40aa-940a-c2fc57ce0f51`

You can look in the network requests to see the initial post request for the conversation, it will be listed under the `plugin_ids` key.

I think an initial low-level implementation of dropping the plugin IDs into the config would be a reasonable start, it would at least allow people to leverage plugins right away. However, I want to make sure the plugin IDs are consistent across users before I take that step.

**Comment by thehunmonkgroup** [comment]
OK, this doesn't look too hard:

There's an endpoint to pull all the plugin info: `https://chat.openai.com/backend-api/aip/p?offset=0&limit=250&statuses=approved`

With entries like this:

```json
{
  "items": [
    {
      "id": "plugin-e5159e19-a6dd-4536-9cad-6a23b82e426e",
      "domain": "voxscript.awt.icu",
      "namespace": "VoxScript",
      "status": "approved",
      "manifest": {
        "schema_version": "v1",
        "name_for_model": "VoxScript",
        "name_for_human": "VoxScript",
        "description_for_model": "Plugin for searching through varius data sources.",
        "description_for_human": "Enables searching of YouTube transcripts, financial data sources, and Google Search results, and more!",
        "auth": {
          "type": "service_http",
          "instructions": "",
          "authorization_type": "bearer",
          "verification_tokens": {
            "openai": "ffc5226d1af346c08a98dee7deec9f76"
          }
        },
        "api": {
          "type": "openapi",
          "url": "https://voxscript.awt.icu/swagger/v1/swagger.yaml"
        },
        "logo_url": "https://voxscript.awt.icu/images/VoxScript_logo_32x32.png",
        "contact_email": "voxscript@allwiretech.com",
        "legal_info_url": "https://voxscript.awt.icu/legal/"
      },
      "oauth_client_id": null,
      "user_settings": {
        "is_installed": false,
        "is_authenticated": true
      },
      "categories": [
        {
          "id": "newly_added",
          "title": "New"
        }
      ]
    },
  ],
}
```

So the plan would be to implement a `/plugin` command that could display a list of plugins and their description to a user. The user would then take the plugin ID of whatever plugin they wanted to enable, and add that to `browser.plugins`. That would inject a `plugin_ids` attribute when a plugin-enabled model (at this point just `gpt-4-plugins`), and from there, things should just work.

**Comment by thehunmonkgroup** [comment]
Implemented and released in 0.10.0
